source/portable/NetworkInterface/DriverSAM/gmac_SAM.c

```diff
@@ -624,29 +624,30 @@ uint32_t gmac_dev_read( gmac_device_t * p_gmac_dev,
         return GMAC_RX_NO_DATA;
     }
 
+    /* Return the number of bytes received. */
+    *p_rcv_size = bytesLeft;
+
     /* gmac_dev_poll has confirmed that there is a complete frame at
      * the current position 'ul_rx_idx'
      */
     nextIdx = p_gmac_dev->ul_rx_idx;
 
-    /* Read +2 bytes because buffers are aligned at -2 bytes */
-    bytesLeft = min( bytesLeft + 2, ( int32_t ) ul_frame_size );
-
     #if ( NETWORK_BUFFERS_CACHED != 0 ) && ( __DCACHE_PRESENT != 0 ) && defined( CONF_BOARD_ENABLE_CACHE )
         SCB_InvalidateDCache();
     #endif
 
     #if ( ipconfigZERO_COPY_RX_DRIVER == 0 )
         {
             /* The frame will be copied in 1 or 2 memcpy's */
-            if( ( p_frame != NULL ) && ( bytesLeft != 0 ) )
+            if( p_frame != NULL )
             {
                 const uint8_t * source;
                 int32_t left;
                 int32_t toCopy;
 
                 source = gs_uc_rx_buffer + nextIdx * GMAC_RX_UNITSIZE;
-                left = bytesLeft;
+                /* Read +2 bytes because buffers are aligned at -2 bytes */
+                left = min( bytesLeft + 2, ( int32_t ) ul_frame_size );
```

Comment:
Note that this code will normally not be executed, only when `ipconfigZERO_COPY_RX_DRIVER = 1`.
We leave it here for those who want to play with the copy-method.
The driver receives frames op to 1514 bytes long. The actual value of `ul_frame_size` is 1536, so the following test is not really necessary:
~~~
left = min( bytesLeft + 2, ( int32_t ) ul_frame_size );
~~~